assets/src/hooks/useDetour.ts

```diff
@@ -75,38 +115,32 @@ export const useDetour = ({ routePatternId, shape }: OriginalRoute) => {
     })
   }
 
-  const canAddWaypoint = () => startPoint !== null && endPoint === null
+  const canAddWaypoint = () =>
+    snapshot.can({
+      type: "detour.edit.place-waypoint",
+      location: { lat: 0, lon: 0 },
+    })
+
   const addWaypoint = canAddWaypoint()
-    ? (p: ShapePoint) => {
-        setWaypoints((positions) => [...positions, p])
+    ? (location: ShapePoint) => {
+        send({ type: "detour.edit.place-waypoint", location })
       }
     : undefined
 
-  const addConnectionPoint = (point: ShapePoint) => {
-    if (startPoint === null) {
-      setStartPoint(point)
-    } else if (endPoint === null) {
-      setEndPoint(point)
-    }
-  }
+  const addConnectionPoint = (point: ShapePoint) =>
+    send({
+      type: "detour.edit.place-connection-point",
+      location: point,
+    })
 
-  const canUndo =
-    startPoint !== null && snapshot.matches({ "Detour Drawing": "Editing" })
+  const canUndo = snapshot.can({ type: "detour.edit.undo" })
```

Comment:
I like this! Feels very state-machiney.

My instincts (admittedly new-ish instincts here) tell me that the main value of using a state machine comes when we otherwise would ask it about its data and variables and state, we instead ask it what it can do, and this feels like a perfect example of that!